instrumentation/google.golang.org/grpc/otelgrpc/test/grpc_stats_handler_test.go

```diff
@@ -1316,3 +1322,75 @@ func checkServerMetrics(t *testing.T, reader metric.Reader) {
 
 	metricdatatest.AssertEqual(t, expectedScopeMetric, rm.ScopeMetrics[0], metricdatatest.IgnoreTimestamp(), metricdatatest.IgnoreValue())
 }
+
+// Ensure there is no data race for the following scenario:
+// Bidirectional streaming + client cancels context in the middle of streaming.
+func TestStatsHandlerConcurrentSafeContextCancellation(t *testing.T) {
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err, "failed to open port")
+	client := newGrpcTest(t, listener,
+		[]grpc.DialOption{
+			grpc.WithStatsHandler(otelgrpc.NewClientHandler()),
+		},
+		[]grpc.ServerOption{
+			grpc.StatsHandler(otelgrpc.NewServerHandler()),
+		},
+	)
+
+	const testCount = 100
+	for i := 0; i < testCount; i++ {
+		t.Run(fmt.Sprintf("run_%d", i), func(t *testing.T) {
+			wg := &sync.WaitGroup{}
+			wg.Add(2)
+			ctx, cancel := context.WithCancel(context.Background())
+			stream, err := client.FullDuplexCall(ctx)
+			require.NoError(t, err)
+			const messageCount = 100
+			go func() {
+				defer wg.Done()
+				sendWg := &sync.WaitGroup{}
+				for i := 0; i < messageCount; i++ {
+					sendWg.Add(1)
+					go func() {
+						defer sendWg.Done()
+						const reqSize = 10
+						pl := interop.ClientNewPayload(testpb.PayloadType_COMPRESSABLE, reqSize)
+						respParam := []*testpb.ResponseParameters{
+							{
+								Size: reqSize,
+							},
+						}
+						req := &testpb.StreamingOutputCallRequest{
+							ResponseType:       testpb.PayloadType_COMPRESSABLE,
+							ResponseParameters: respParam,
+							Payload:            pl,
+						}
+						err := stream.Send(req)
+						if err == io.EOF { // possible due to context cancellation
+							require.ErrorIs(t, ctx.Err(), context.Canceled)
+						} else {
+							require.NoError(t, err)
+						}
+					}()
+				}
+				sendWg.Wait()
+				require.NoError(t, stream.CloseSend())
+			}()
+			go func() {
+				defer wg.Done()
+				for i := 0; i < messageCount; i++ {
+					_, err := stream.Recv()
+					if i > messageCount/2 {
+						cancel()
+					}
+					// must continue to receive messages until server acknowledges the cancellation, to ensure no data race happens there too
+					if status.Code(err) == codes.Canceled {
+						return
+					}
+					require.NoError(t, err)
+				}
+			}()
+			wg.Wait()
+		})
```

Comment:
I just propose to simplify the test:

1. We do not need to (and even should not) call each `err := stream.Send(req)` in a separate goroutine to reproduce the issue. Take notice that it is an unsafe usage of grpc. From https://pkg.go.dev/google.golang.org/grpc#ClientStream: "is not safe to call SendMsg on the same stream in different goroutines".
2. I do not see any benefit of using `t.Run` in this scenario
3. `testCount = 100` changed to `n = 10`, `messageCount` to `10` (no need to have so many iterations to repro the race)
4. Changed `reqSize` to `1`
5. Refactored the use of `wg sync.WaitGroup`


```suggestion
	const n = 10
	for i := 0; i < n; i++ {
		ctx, cancel := context.WithCancel(context.Background())
		stream, err := client.FullDuplexCall(ctx)
		require.NoError(t, err)

		const messageCount = 10
		var wg sync.WaitGroup

		wg.Add(1)
		go func() {
			defer wg.Done()
			for i := 0; i < messageCount; i++ {
				const reqSize = 1
				pl := interop.ClientNewPayload(testpb.PayloadType_COMPRESSABLE, reqSize)
				respParam := []*testpb.ResponseParameters{
					{
						Size: reqSize,
					},
				}
				req := &testpb.StreamingOutputCallRequest{
					ResponseType:       testpb.PayloadType_COMPRESSABLE,
					ResponseParameters: respParam,
					Payload:            pl,
				}
				err := stream.Send(req)
				if err == io.EOF { // possible due to context cancellation
					require.ErrorIs(t, ctx.Err(), context.Canceled)
				} else {
					require.NoError(t, err)
				}
			}
			require.NoError(t, stream.CloseSend())
		}()

		wg.Add(1)
		go func() {
			defer wg.Done()
			for i := 0; i < messageCount; i++ {
				_, err := stream.Recv()
				if i > messageCount/2 {
					cancel()
				}
				// must continue to receive messages until server acknowledges the cancellation, to ensure no data race happens there too
				if status.Code(err) == codes.Canceled {
					return
				}
				require.NoError(t, err)
			}
		}()

		wg.Wait()
```